### DISCUSSION TITLE
InitialScript failed Postgres container startup

### DISCUSSION THREAD

**Discussion by hantsy** [testcontainers/testcontainers-java#4841] [Q&A] [answered]
I have an initial script like this.

```sql
DO $$ BEGIN
    CREATE TYPE post_status AS ENUM( 'DRAFT', 'PENDING_MODERATION', 'PUBLISHED');
EXCEPTION
    WHEN duplicate_object THEN null;
END $$;
```
It worked well when using `docker compose` command.

But when I create a Postgres docker container using testcontainers like this.

```java
 var container = new PostgreSQLContainer("postgres:12")
                    .withInitScript("init.sql");
```
It will fail with the following exception.

```bash 
Caused by: org.postgresql.util.PSQLException: Unterminated dollar quote started at position 3 in SQL DO $$ BEGIN CREATE TYPE post_status AS ENUM( 'DRAFT', 'PENDING_MODERATION', 'PUBLISHED'). Expected terminating $$
```

The example project to reproduce this issue: https://github.com/hantsy/spring6-sandbox/blob/master/jdbc/src/test/java/com/example/demo/testcontainers/PostRepositoryTestWithTestcontainers.java

**Comment by kiview** [comment, marked as answer]
Hey @hantsy, this is related to this issue:
https://github.com/testcontainers/testcontainers-java/issues/4441

The best solution is to delegate the script execution to Postgres itself. This can be achieved by copying the file into a special location in the container (a feature of the Postgres image):
```java
new PostgreSQLContainer("postgres:12")
                    .withCopyFileToContainer(
                        MountableFile.forClasspathResource("init.sql"), 
                        "/docker-entrypoint-initdb.d/init.sql"
                     );
```

**Reply by seanf** [reply]
@kiview How can I do this with [org.testcontainers.jdbc.ContainerDatabaseDriver](https://www.testcontainers.org/modules/databases/jdbc/) please? 

I want to provide jooq-codegen-maven with a pre-initialised database. This was working until the functions in the schema started using `$$`.

**Reply by seanf** [reply]
> @kiview How can I do this with [org.testcontainers.jdbc.ContainerDatabaseDriver](https://www.testcontainers.org/modules/databases/jdbc/) please?
> 
> I want to provide jooq-codegen-maven with a pre-initialised database. This was working until the functions in the schema started using `$$`.

I found a workaround for Maven/jOOQ, but it's got a lot of moving parts.

It requires an extra Maven module. This module holds one class to implement an init function. The init function reads in the init.sql file as one big string, then executes it via JDBC's `Statement.execute()`. That module is declared as a dependency of the jooq-codegen-maven plugin, and the JDBC URL uses the `TC_INITFUNCTION` parameter to point to the init function. Fiddly, but it works.

**Reply by kiview** [reply]
I get the complexity @seanf. For this to become simpler using the special JDBC URL integration, we probably would need to have a feature like a URL query variable, that delegates to `withCopyFileToContainer`. Since the location will be specific to the db image, this could be somehow implemented in the specific container implementation.
